rust/src/applayer.rs

```diff
@@ -507,3 +507,56 @@ pub unsafe fn get_event_info_by_id<T: AppLayerEvent>(
     }
     return -1;
 }
+
+/// Transaction trait.
+///
+/// This trait defines methods that a Transaction struct must implement
+/// in order to define some generic helper functions.
+pub trait Transaction {
+    fn id(&self) -> u64;
+
+    /// Get the app-layer decoder events from the transaction.
+    ///
+    /// A default implementation is provided for app-layer implememtations
+    /// that do not have decoder events.
+    fn get_events(&self) -> *mut AppLayerDecoderEvents {
+        std::ptr::null_mut()
+    }
+}
+
+pub trait State<Tx: Transaction> {
+    fn get_transactions(&self) -> &[Tx];
+
+    fn get_transaction_iterator(&self, min_tx_id: u64, state: &mut u64) -> AppLayerGetTxIterTuple {
+        let mut index = *state as usize;
+        let transactions = self.get_transactions();
+        let len = transactions.len();
+        while index < len {
+            let tx = &transactions[index];
+            if tx.id() < min_tx_id + 1 {
```

Comment:
Is there a cost to calling a function `tx.id()` ?
The other alternative I see to using traits are having macros. What do you think about this choice trait versus macros ?

Reply:
`tx.id()` has no cost when build in release mode, simple getters like this are optimized away. You do see the `call` in the disassembled code when built in debug mode.

I think generics like this are preferable to macros as they are proper code that can be read, refactored, type checked, etc.  In this case as no dynamic dispatch occurrs, its actiing more or less like a macro.

Reply:
> tx.id() has no cost when build in release mode, simple getters like this are optimized away.

That is awesome